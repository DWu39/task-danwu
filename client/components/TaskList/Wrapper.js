import React from 'react';
import styled from 'styled-components';

import theme from '../../style/theme';

const Wrapper = styled.section`
  background: ${theme.white};
  border-radius: 4px;
  box-shadow: 0 1px 1px 0 ${theme.gray};
  box-sizing: border-box;
  height: 180px;
  margin-top: 14px;
`;

export default Wrapper;
