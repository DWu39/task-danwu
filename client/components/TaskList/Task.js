import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';

import Button from '../Button';
import Wrapper from './Wrapper';
import theme from '../../style/theme';

const StyledTask = Wrapper.extend`
  padding: 14px;

  display: flex;
  align-items: flex-start;
  justify-content: space-between;
`;

const StyledTaskText = styled.p`
  color: ${theme.gray};
  margin: 0;
  text-transform: uppercase;
`;

const TrashButton = styled(Button)`
  background: ${theme.white};
  color: ${theme.gray};
  font-size: 16px;
  padding: 0;
`;

const handleInputChange = (handler) => {
  return (event) => {
    handler(event.target.value);
  };
}

const handleKeyDown = (handler) => {
  return (event) => {
    // pressed enter, escape, or tab
    if (event.key === 'Enter' || event.keyCode === 27 || event.keyCode === 9) {
      event.preventDefault();
      handler();
    }
  };
}

const Task = ({
  closeTask,
  deleteTask,
  editTask,
  openTask,
  isOpened,
  text
}) => (
  <StyledTask>
    {isOpened
      ? <input
          autoFocus
          onBlur={closeTask}
          onChange={handleInputChange(editTask)}
          onKeyDown={handleKeyDown(closeTask)}
          placeholder='Edit task description'
          type='text'
        />
      : <StyledTaskText
          onClick={openTask}
          onFocus={openTask}
          tabIndex={0}
        >
          {text}
        </StyledTaskText>
    }
    <TrashButton onClick={deleteTask}>
      <i className="fa fa-trash-o"></i>
    </TrashButton>
  </StyledTask>
);

Task.propTypes = {
  deleteTask: PropTypes.func,
  editTask: PropTypes.func,
  openTask: PropTypes.func,
  isOpened: PropTypes.bool,
  text: PropTypes.string
}

Task.defaultProps = {
  deleteTask: null,
  editTask: null,
  openTask: null,
  isOpened: false,
  text: PropTypes.string
}

export default Task;
