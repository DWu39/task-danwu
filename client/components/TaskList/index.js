import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';

import Task from './Task';
import Wrapper from './Wrapper';
import theme from '../../style/theme';

const StyledTaskList = styled.ul`
  margin: 0;
  margin-bottom: 100px;
  padding: 0;
`;

const EmptyTaskList = Wrapper.extend`
  color: ${theme.navy};
  font-size: 20px;

  display: flex;
  flex-direction:column;
  justify-content:center;
  text-align: center;
`;

const TaskList = ({
  closeTask,
  deleteTask,
  openTask,
  tasks
}) => (
  <StyledTaskList>
    {tasks.length
      ? tasks.map((task) => (
        <Task
          key={task.id}
          closeTask={(text) => { closeTask(task.id, text); }}
          deleteTask={() => { deleteTask(task.id); }}
          openTask={() => { openTask(task.id); }}
          isOpened={task.isOpened}
          text={task.text}
        />
      ))
      : <EmptyTaskList>You don't have any tasks. Start by clicking "Add Task"!</EmptyTaskList>
    }
  </StyledTaskList>
);

TaskList.propTypes = {
  closeTask: PropTypes.func,
  deleteTask: PropTypes.func,
  openTask: PropTypes.func,
  tasks: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired
    })
  )
}

TaskList.defaultProps = {
  closeTask: null,
  deleteTask: null,
  openTask: null,
  tasks: []
}

export default TaskList;
