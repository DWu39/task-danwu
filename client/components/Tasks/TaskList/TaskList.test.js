import React from 'react';
import { shallow } from 'enzyme';

import TaskList from './index';

const mockTasks = [
  {
    id: 'id1',
    text: 'Buy milk'
  },
  {
    id: 'id2',
    text: 'Walk the dog'
  },
  {
    id: 'id3',
    text: 'Task'
  }
];

describe('TaskList', () => {
  it('should render empty correctly', () => {
    const output = shallow(<TaskList/>);
    expect(output).toMatchSnapshot();
  });

  it('should render tasks correctly', () => {
    const output = shallow(<TaskList tasks={mockTasks}/>);
    expect(output).toMatchSnapshot();
  });
});
