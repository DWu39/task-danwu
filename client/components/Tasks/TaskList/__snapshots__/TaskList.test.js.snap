// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList should render empty correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskList
    closeTask={null}
    deleteTask={null}
    openTask={null}
    tasks={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <styled.section>
        You don't have any tasks. Start by clicking "Add Task"!
      </styled.section>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "You don't have any tasks. Start by clicking \\"Add Task\\"!",
      },
      "ref": null,
      "rendered": "You don't have any tasks. Start by clicking \\"Add Task\\"!",
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <styled.section>
          You don't have any tasks. Start by clicking "Add Task"!
        </styled.section>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "You don't have any tasks. Start by clicking \\"Add Task\\"!",
        },
        "ref": null,
        "rendered": "You don't have any tasks. Start by clicking \\"Add Task\\"!",
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`TaskList should render tasks correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskList
    closeTask={null}
    deleteTask={null}
    openTask={null}
    tasks={
      Array [
        Object {
          "id": "id1",
          "text": "Buy milk",
        },
        Object {
          "id": "id2",
          "text": "Walk the dog",
        },
        Object {
          "id": "id3",
          "text": "Task",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Task
          closeTask={[Function]}
          deleteTask={[Function]}
          isOpened={false}
          openTask={[Function]}
          text="Buy milk"
        />,
        <Task
          closeTask={[Function]}
          deleteTask={[Function]}
          isOpened={false}
          openTask={[Function]}
          text="Walk the dog"
        />,
        <Task
          closeTask={[Function]}
          deleteTask={[Function]}
          isOpened={false}
          openTask={[Function]}
          text="Task"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "id1",
        "nodeType": "function",
        "props": Object {
          "closeTask": [Function],
          "deleteTask": [Function],
          "isOpened": false,
          "openTask": [Function],
          "text": "Buy milk",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "id2",
        "nodeType": "function",
        "props": Object {
          "closeTask": [Function],
          "deleteTask": [Function],
          "isOpened": false,
          "openTask": [Function],
          "text": "Walk the dog",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "id3",
        "nodeType": "function",
        "props": Object {
          "closeTask": [Function],
          "deleteTask": [Function],
          "isOpened": false,
          "openTask": [Function],
          "text": "Task",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Task
            closeTask={[Function]}
            deleteTask={[Function]}
            isOpened={false}
            openTask={[Function]}
            text="Buy milk"
          />,
          <Task
            closeTask={[Function]}
            deleteTask={[Function]}
            isOpened={false}
            openTask={[Function]}
            text="Walk the dog"
          />,
          <Task
            closeTask={[Function]}
            deleteTask={[Function]}
            isOpened={false}
            openTask={[Function]}
            text="Task"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "id1",
          "nodeType": "function",
          "props": Object {
            "closeTask": [Function],
            "deleteTask": [Function],
            "isOpened": false,
            "openTask": [Function],
            "text": "Buy milk",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "id2",
          "nodeType": "function",
          "props": Object {
            "closeTask": [Function],
            "deleteTask": [Function],
            "isOpened": false,
            "openTask": [Function],
            "text": "Walk the dog",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "id3",
          "nodeType": "function",
          "props": Object {
            "closeTask": [Function],
            "deleteTask": [Function],
            "isOpened": false,
            "openTask": [Function],
            "text": "Task",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
